uniffi_core/src/ffi/rustfuture.rs

```diff
@@ -115,477 +75,638 @@
 //! [`Waker`]: https://doc.rust-lang.org/std/task/struct.Waker.html
 //! [`RawWaker`]: https://doc.rust-lang.org/std/task/struct.RawWaker.html
 
-use crate::{
-    ffi::foreignexecutor::RustTaskCallbackCode, rust_call_with_out_status, schedule_raw,
-    FfiConverter, FfiDefault, ForeignExecutor, ForeignExecutorHandle, RustCallStatus,
-};
 use std::{
-    cell::UnsafeCell,
     future::Future,
+    marker::PhantomData,
+    mem,
+    ops::Deref,
     panic,
     pin::Pin,
-    sync::{
-        atomic::{AtomicU32, Ordering},
-        Arc,
-    },
-    task::{Context, Poll, RawWaker, RawWakerVTable, Waker},
+    sync::{Arc, Mutex},
+    task::{Context, Poll, Wake},
 };
 
-/// Callback that we invoke when a `RustFuture` is ready.
+use once_cell::sync::OnceCell;
+
+use crate::{rust_call_with_out_status, FfiConverter, FfiDefault, RustCallStatus};
+
+/// Result code for [rust_future_poll].  This is passed to the continuation function.
+#[repr(i8)]
+#[derive(Debug, PartialEq, Eq)]
+pub enum RustFuturePoll {
+    /// The future is ready and is waiting for [rust_future_complete] to be called
+    Ready = 0,
+    /// The future might be ready and [rust_future_poll] should be called again
+    MaybeReady = 1,
+}
+
+/// Foreign callback that's passed to [rust_future_poll]
 ///
-/// The foreign code passes a pointer to one of these callbacks along with an opaque data pointer.
-/// When the future is ready, we invoke the callback.
-pub type FutureCallback<T> =
-    extern "C" fn(callback_data: *const (), result: T, status: RustCallStatus);
+/// The Rust side of things calls this when the foreign side should call [rust_future_poll] again
+/// to continue progress on the future.
+pub type RustFutureContinuationCallback = extern "C" fn(callback_data: *const (), RustFuturePoll);
+
+/// Opaque handle for a Rust future that's stored by the foreign language code
+#[repr(transparent)]
+pub struct RustFutureHandle(*const ());
+
+/// Stores the global continuation callback
+static RUST_FUTURE_CONTINUATION_CALLBACK_CELL: OnceCell<RustFutureContinuationCallback> =
+    OnceCell::new();
+
+/// Set the global RustFutureContinuationCallback.
+pub fn rust_future_continuation_callback_set(callback: RustFutureContinuationCallback) {
+    if let Err(existing) = RUST_FUTURE_CONTINUATION_CALLBACK_CELL.set(callback) {
+        // Don't panic if this to be called multiple times with the same callback.
+        if existing != callback {
+            panic!("Attempt to set the RustFuture continuation callback twice");
+        }
+    }
+}
```

Comment:
I thought it would be a lot harder to fix the miri issues, but once I realized `OnceCell` exists it was easy.  I think we should eventually replace the other callback cell code to use it too.